drivers/modem/ublox-sara-n310.c

```diff
@@ -0,0 +1,526 @@
+/*
+ * Copyright (c) 2021 Abel Sensors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+#include <net/net_if.h>
+#include <net/net_offload.h>
+#include <net/socket_offload.h>
+
+#include <kernel.h>
+#include <ctype.h>
+#include <errno.h>
+#include <zephyr.h>
+#include <drivers/gpio.h>
+#include <device.h>
+#include <init.h>
+
+#include <stdio.h>
+
+#include <drivers/modem/ublox-sara-n310.h>
+
+#include "modem_context.h"
+#include "modem_iface_uart.h"
+#include "modem_cmd_handler.h"
+
+#define MAX_DATA_LENGTH 1024
+#define RX_PRIORITY 5
+
+#define MDM_RECV_MAX_BUF 10
+#define MDM_RECV_BUF_SIZE 64
+
+#define DT_DRV_COMPAT ublox_saran310
+
+LOG_MODULE_REGISTER(modem_ublox_saran310, CONFIG_MODEM_LOG_LEVEL);
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+
+/* modem info */
+struct modem_info {
+	char mdm_iccid[64];
+	char mdm_model[64];
+	char mdm_manufacturer[64];
+	char mdm_revision[64];
+	char mdm_imei[64];
+
+	char mdm_ip[64];
+};
+
+static struct modem_info minfo;
+
+/* modem_info getters */
+char *n310_get_model(void)
+{
+	return minfo.mdm_model;
+}
+
+char *n310_get_iccid(void)
+{
+	return minfo.mdm_iccid;
+}
+
+char *n310_get_manufacturer(void)
+{
+	return minfo.mdm_manufacturer;
+}
+
+char *n310_get_revision(void)
+{
+	return minfo.mdm_revision;
+}
+
+char *n310_get_imei(void)
+{
+	return minfo.mdm_imei;
+}
+
+char *n310_get_ip(void)
+{
+	return minfo.mdm_ip;
+}
+
+/* data struct */
+struct modem_data {
+	struct modem_context context;
+	struct modem_cmd_handler_data cmd_handler_data;
+	struct modem_iface_uart_data mdm_data;
+
+	uint8_t cmd_match_buf[128];
+	char mdm_rx_rb_buf[1024];
+
+	struct k_sem sem_response;
+
+	enum n310_network_state network_state;
+};
+
+static struct modem_data mdata;
+
+/* rx thread data */
+NET_BUF_POOL_DEFINE(mdm_recv_pool, MDM_RECV_MAX_BUF, MDM_RECV_BUF_SIZE, 0,
+		    NULL);
+K_THREAD_STACK_DEFINE(modem_rx_stack, MAX_DATA_LENGTH);
+static struct k_thread rx_data;
+
+/* AT command struct */
+static struct AT {
+	uint8_t *CCID;
+	uint8_t *NSOSTF;
+	uint8_t *USORF;
+} at;
+
+/* check if module has IP address */
+static bool has_ip(void)
+{
+	return !(minfo.mdm_ip == NULL);
+}
+
+/* get network state */
+int n310_get_state(void)
+{
+	return mdata.network_state;
+}
+
+/* AT command handler functions */
+MODEM_CMD_DEFINE(on_cmd_ok)
+{
+	modem_cmd_handler_set_error(data, 0);
+	k_sem_give(&mdata.sem_response);
+	return 0;
+}
+
+MODEM_CMD_DEFINE(on_cmd_error)
+{
+	modem_cmd_handler_set_error(data, -EINVAL);
+	k_sem_give(&mdata.sem_response);
+	return 0;
+}
+
+MODEM_CMD_DEFINE(on_cmd_socknotifycereg)
+{
+	size_t out_len;
+	char buf[64];
+
+	out_len = net_buf_linearize(buf, sizeof(buf) - 1, data->rx_buf, 0, len);
+	buf[out_len] = '\0';
+
+	mdata.network_state = buf[0] - '0';
```

Comment:
I'm not sure it is allowed to subtract a character. See rule 10.2 in https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html